internal/profiling/profiling.go

```diff
@@ -0,0 +1,251 @@
+func IsEnabled() bool {
+	return atomic.LoadUint32(&profilingEnabled) > 0
+}
+
+// Enable turns profiling on and off.
+//
+// Note that it is impossible to enable profiling for one server and leave it
+// turned off for another. This is intentional and by design -- if the status
+// of profiling was server-specific, clients wouldn't be able to profile
+// themselves. As a result, SetEnabled turns profiling on and off for all
+// servers and clients in the binary. Each stat will be, however, tagged with
+// whether it's a client stat or a server stat; so you should be able to filter
+// for the right type of stats in post-processing.
+// SetEnabled is the internal
+func Enable(enabled bool) {
+	if enabled {
+		atomic.StoreUint32(&profilingEnabled, 1)
+	} else {
+		atomic.StoreUint32(&profilingEnabled, 0)
+	}
+}
+
+// A Timer represents the wall-clock beginning and ending of a logical
+// operation.
+type Timer struct {
+	// Each Timer has a tag that identifies it uniquely within a Stat. This is
+	// usually a forward-slash-separated hierarchical string.
+	TimerTag string
+	// Begin marks the beginning of this timer. The timezone is unspecified, but
+	// must use the same timezone as End; this is so shave off the small, but
+	// non-zero time required to convert to a standard timezone such as UTC.
+	Begin time.Time
+	// End marks the end of a timer.
+	End time.Time
+	// Each Timer must be started and ended within the same goroutine; GoID
+	// captures this goroutine ID. The Go runtime doesn not typically expose this
+	// information, so this is set to zero in the typical case. However, a
+	// trivial patch to the runtime package can make this field useful. See
+	// goid_modified.go in this package for more details.
+	GoID int64
+}
+
+// NewTimer creates and returns a new Timer object. This is useful when you
+// don't already have a Stat object to associate this Timer with; for example,
+// before the context of a new RPC query is created, a Timer may be needed to
+// measure transport-related operations.
+//
+// Use AppendTimer to append the returned Timer to a Stat.
+func NewTimer(timerTag string) *Timer {
+	return &Timer{
+		TimerTag: timerTag,
+		Begin:    time.Now(),
+		GoID:     goid(),
+	}
+}
+
+// A Stat is a collection of Timers that represent timing information for
+// different components within this Stat. For example, a Stat may be used to
+// reference the entire lifetime of an RPC request, with Timers within it
+// representing different components such as encoding, compression, and
+// transport.
+//
+// The user is expected to use the included helper functions to do operations
+// on the Stat such as creating or appending a new timer. Direct operations on
+// the Stat's exported fields (which are exported for encoding reasons) may
+// lead to data races.
+type Stat struct {
+	// Each Stat has a tag associated with it to identify it in post-processing.
+	// This is can be any arbitrary string.
+	StatTag string
```

Comment:
Same as timer, just `Tag`.
We can keep it if renaming isn't trivial.

Reply:
Done.